# Part 1. Equation of a Slime

How many late days are you using for this assignment?  1

In [28]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
import numpy as np
from sklearn.preprocessing import PolynomialFeatures

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler



## 1. Loading the dataset

In [29]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
data = pd.read_csv('science_data_large.csv')
print(data.head(15))
data.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [30]:
# Split into features (X) and label (y)
X = data[['Temperature °C', 'Mols KCL']]
y = data['Size nm^3']

# Split into training (90%) and testing (10%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [31]:
# Step 1: Split the dataset into features (X) and label (y)
X = data[['Temperature °C', 'Mols KCL']] # Ensuring feature names are preserved
y = data['Size nm^3']

# Step 2: Split into training (90%) and testing (10%) datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# Step 3: Use sklearn to train a linear regression model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Step 4: Create a sample datapoint and predict the output
sample_data = pd.DataFrame([[500, 800]], columns=['Temperature °C', 'Mols KCL'])  # Create DataFrame with column names

predicted_size = model.predict(sample_data)
print("Predicted Size nm^3 for sample data:", predicted_size[0])  # Print the predicted size

# Step 5: Report the R² score for the model
score = model.score(X_test, y_test)
print("Model R² score:", score)

Predicted Size nm^3 for sample data: 849837.7802911897
Model R² score: 0.8552472077276095


Write the linear equation of a slime: (example equation: $E = mc^2$)

Size (nm³) = Intercept + Coefficient_Temperature * Temperature + Coefficient_KCl * Mols KCl

Size (nm³) = -409391.48 + 866.15 * Temperature + 1032.70 * Mols KCl


Report on score and explain meaning: The R² score, known as the coefficient of determination, indicates how well the independent variables explain the variance in the dependent variable. 
A score of 1.0 indicates a perfect fit, while a score of 0.0 indicates no explanatory power. 
A higher R² score implies a better model performance in predicting the outcome variable.

## 4. Use Cross Validation

In [32]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
cross_val_scores = cross_val_score(model, X, y, cv=5)
print("Cross-validation scores:", cross_val_scores)
print("Mean cross-validation score:", np.mean(cross_val_scores))
# Report on their finding and their significance

Cross-validation scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean cross-validation score: 0.8568167899144437


Write findings here: Overall cross-validation findings indicate that the linear regression model is performing well since the cross-validation score is high. It consistently captures a large portion of the variance in the data, suggesting that the chosen features have strong predictive power for determining slime size.

## 5. Using Polynomial Regression

In [33]:
# Create polynomial features with degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

# Fit linear regression model on polynomial features
poly_model = LinearRegression()
poly_model.fit(X_poly, y)

# Extract and display coefficients and intercept
coefficients = poly_model.coef_
intercept = poly_model.intercept_
terms = poly.get_feature_names_out(['Temperature °C', 'Mols KCL'])

# Perform 5-fold cross-validation on the polynomial model
poly_cv_scores = cross_val_score(poly_model, X_poly, y, cv=5)
print("Polynomial Regression Cross-validation scores:", poly_cv_scores)
print("Mean cross-validation score:", np.mean(poly_cv_scores))

Polynomial Regression Cross-validation scores: [1. 1. 1. 1. 1.]
Mean cross-validation score: 1.0


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Size (nm³) = 2e-05 + (12.0)*Temperature °C + (-0.0)*Mols KCL + (-0.0)*Temperature °C^2 + (2.0)*Temperature °C Mols KCL + (0.02857)*Mols KCL^2

Report on the score and interpret: A score of 1.0 implies that the model explains all variance in the data for the validation sets, which is very rare (and is not super possible) in practical scenarios and suggests that the model may be overfitting the data, especially if your dataset is not very large or diverse.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [34]:
# Load the dataset. Then train and evaluate the classification models.
ckd_data = pd.read_csv('ckd_feature_subset.csv')
print(ckd_data.head(15))
ckd_data.info()

         age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0   0.688312  0.333333  0.000000           1           0  0.000000           1
1   0.545455  0.333333  0.128319           1           0  0.305085           1
2   0.714286  0.500000  0.238938           1           0  0.186441           1
3   0.688312  0.333333  0.283186           0           1  0.338983           1
4   0.441558  0.333333  0.221239           1           0  0.220339           1
5   0.740260  0.833333  0.265487           0           1  0.355932           1
6   0.870130  0.333333  0.668142           0           1  0.237288           1
7   0.870130  0.500000  0.150442           0           1  0.372881           1
8   0.194805  0.666667  0.380531           0           1  0.305085           1
9   0.753247  0.833333  0.163717           0           1  0.220339           1
10  0.701299  0.166667  0.504425           1           0  0.152542           1
11  0.610390  0.666667  0.367257           1        

## Results and Conclusion for Classification Experiments

In [37]:
# Step 1: Split into features and label
X_ckd = ckd_data.drop(columns=['Target_ckd'])
y_ckd = ckd_data['Target_ckd']

# Step 2: Use a fixed random seed of 42
X_train_ckd, X_test_ckd, y_train_ckd, y_test_ckd = train_test_split(X_ckd, y_ckd, test_size=0.1, random_state=42)

# Step 3: Scale the features
scaler = StandardScaler()
X_train_ckd = scaler.fit_transform(X_train_ckd)
X_test_ckd = scaler.transform(X_test_ckd)

# Step 4: Define the models
models = {
    'Logistic Regression': LogisticRegression(max_iter=2500),  # Increased max_iter
    'SVC': SVC(),
    'KNeighborsClassifier': KNeighborsClassifier(),
    'MLPClassifier': MLPClassifier(max_iter=2500)  # Increased max_iter
}

results = {}

# Step 5: Evaluate the models using cross-validation
for name, model in models.items():
    cv_scores = cross_val_score(model, X_train_ckd, y_train_ckd, cv=5)
    results[name] = {
        'mean_score': np.mean(cv_scores),
        'std_score': np.std(cv_scores)
    }
    print(f"{name} Cross-validation scores: {cv_scores}")

# Step 6: Store results in a DataFrame
results_df = pd.DataFrame(results).T
print(results_df)

# Step 7: Example of exploring different configurations of the MLPClassifier
nn_configs = [
    {'hidden_layer_sizes': (5,), 'activation': 'tanh'},
    {'hidden_layer_sizes': (10,), 'activation': 'relu'},
    {'hidden_layer_sizes': (15, 5), 'activation': 'logistic'}  # This is acceptable if 'logistic' is recognized
]

nn_results = {}

# Step 8: Evaluate different neural network configurations
for config in nn_configs:
    model = MLPClassifier(hidden_layer_sizes=config['hidden_layer_sizes'], 
                          activation=config['activation'], 
                          max_iter=2500, 
                          verbose=False) 
    cv_scores = cross_val_score(model, X_train_ckd, y_train_ckd, cv=5)
    nn_results[f"{config['hidden_layer_sizes']}, {config['activation']}"] = {
        'mean_score': np.mean(cv_scores),
        'std_score': np.std(cv_scores)
    }

# Step 9: Store neural network results in a DataFrame
nn_results_df = pd.DataFrame(nn_results).T
print(nn_results_df)

Logistic Regression Cross-validation scores: [0.92857143 0.96428571 0.88888889 0.88888889 0.96296296]
SVC Cross-validation scores: [1.         1.         0.96296296 0.88888889 0.96296296]
KNeighborsClassifier Cross-validation scores: [0.96428571 0.92857143 0.85185185 0.77777778 0.96296296]
MLPClassifier Cross-validation scores: [1.         1.         0.92592593 0.88888889 0.96296296]
                      mean_score  std_score
Logistic Regression     0.926720   0.033438
SVC                     0.962963   0.040572
KNeighborsClassifier    0.897090   0.072288
MLPClassifier           0.955556   0.043192
                   mean_score  std_score
(5,), tanh           0.941005   0.044580
(10,), relu          0.933598   0.054579
(15, 5), logistic    0.963228   0.040575


## Results and Conclusion for Neural Network Experiments

The evaluation of various classification models on the Chronic Kidney Disease (CKD) dataset yielded insightful results regarding their performance. Starting with Logistic Regression, this model achieved cross-validation scores ranging from approximately 0.889 to 0.964, which culminated in a mean score of 0.926720 with a standard deviation of 0.033438. These results indicate that Logistic Regression can effectively discriminate between classes, although the variability shows it may struggle with particular subsets of the data. The Support Vector Classifier (SVC) outperformed the other models, achieving perfect scores in three of the five folds. This resulted in a mean score of 0.962963 and a standard deviation of 0.040572, suggesting a robust ability to classify CKD instances accurately.


Additionally, the KNeighbors Classifier exhibited more variability in its scores, with a mean of 0.897090 and a higher standard deviation of 0.072288, indicating potential sensitivity to the choice of neighbors and the presence of noise in the data. Lastly, the MLP Classifier demonstrated strong performance across the board, with cross-validation scores averaging 0.955556 and a standard deviation of 0.043192. Notably, while the MLP Classifier's performance was commendable, it is important to consider the configurations tested for hyperparameter tuning. The configurations with different hidden layers yielded several interesting results; the MLP with 15 hidden nodes and logistic activation reached a mean score of 0.963228. This indicates that while neural networks can adequately capture complex patterns in the data, the choice of architecture significantly impacts performance stability. Overall, the SVC and MLP models emerged as the most effective for this classification task, with logistic regression providing a strong baseline performance. Further tuning, especially of the MLP Classifier, may reveal even better model fits for the CKD dataset.